README.md
# Chatbot Microservice

This repository contains a microservice for a chatbot application. The microservice is built using FastAPI and integrates with the OpenAI API for natural language understanding. It allows users to interact with the chatbot by sending text messages and receiving responses.


## Features

- Accepts user input text and provides responses using OpenAI API.
- Stores conversation history in MongoDB.
- Provides an API for extracting the conversation history.
- Integrates with FastAPI for the web service.


## Prerequisites

- Python 3.7 or later
- [pip](https://pip.pypa.io/en/stable/installing/) package manager
- [MongoDB](https://docs.mongodb.com/manual/installation/) database
- [OpenAI API key](https://beta.openai.com/docs/developer-quickstart/your-api-keys)

## Directory Structure

The repository has the following structure:

```
chatbot-microservice/
  |- outputs/                     # Sample outputs folder
  |   |- history.txt              # History text of the given user
  |   |- output.mp3                 # Sample response audio file
  |- start_bot.py          # Python script to start and use the bot
  |- microservice.py          # Python script for the chatbot micro service
  |- microservice.ipynb          # Python notebook to test and use the micro service chatbot
  |- cbtbox.html               # HTML file for the chat UI
  |- requirements.txt         # List of required Python packages
  |- README.md                # This README file
```
##

## Files Explanation

- `start_bot.py`: This is a Python script used to start and use the bot.
  
- `microservice.py`: This Python script creates the chatbot microservice using the FastAPI framework. It defines the endpoints for chatting with the bot, getting the history of a user, and fetching the latest audio response. The script also includes the logic for interacting with the MongoDB database and OpenAI's GPT-3.5 model.

- `microservice.ipynb`: This Jupyter notebook demonstrates how to test and use the microservice chatbot. It includes making HTTP requests to the microservice and handling the responses.

- `chatbox.html`: This HTML file is the frontend chat UI that interacts with the chatbot microservice. Users can send messages and receive responses from the bot in this interface.


## Setup

1. Clone the repository:

   ```bash
   git clone https://github.com/your-username/chatbot-microservice.git
   cd chatbot-microservice
   ```

2. Create and activate a virtual environment (requires Python 3.7+):

   ```bash
   python3 -m venv .venv
   source .venv/bin/activate
   ```

3. Install the required packages:

   ```bash
   pip install -r requirements.txt
   ```
## Usage 1

To run the microservice and interact with the bot using the Swagger UI, use the `uvicorn` ASGI server:

1. Ensure that your virtual environment is active, and you have installed all necessary dependencies by following the steps in the Setup section.

2. Then run this command `$env:MONGO_CONNECTION_STRING="mongodb://localhost:27017` to set the environment variable for the MongoDB connection string.

3. Next run the command:

   ```bash
   uvicorn microservice:app --reload
   ```

   This command starts the `uvicorn` server with the FastAPI application defined in `microservice.py`. The `--reload` flag enables hot reloading, which means the server will automatically update whenever you make changes to the code.

3. Open your browser and navigate to the base URL [http://127.0.0.1:8000](http://127.0.0.1:8000). You should see a message confirming that the chatbot is running.

4. To interact with the chatbot and test the API endpoints, navigate to [http://127.0.0.1:8000/docs](http://127.0.0.1:8000/docs). This URL opens the interactive FastAPI documentation generated by Swagger UI. Here you can:

   - Send POST requests to the `/chat` endpoint to chat with the bot. You will need to provide a `user_id` and a `message` in the request body.
   - Send GET requests to the `/history/{user_id}` endpoint to retrieve the conversation history for a specific user. Replace `{user_id}` in the URL with the actual user ID.
   - Send GET requests to the `/audio` endpoint to retrieve the latest response from the bot as an audio file.

### Sample UI 1
![Sample UI1](fastapi1.PNG)
![Sample UI2](fastapi2.PNG)
## Usage 2

Use `microservice.ipynb` to test and interact with the microservice chatbot. 

### Sample UI 2
![Sample UI3](UI.PNG)






outputs/history.txt
User: What is earth?
Assistant: Earth is the third planet from the Sun in our solar system. It is the largest and densest of the four rocky planets, which also include Mercury, Venus, and Mars. Earth is unique in our solar system because it is the only known planet to support life. It has a diverse range of ecosystems, including bodies of water, landmasses, and a breathable atmosphere. Earth is also the home to millions of species, including humans.

start_bot.py
import subprocess
import http.server
import socketserver
import webbrowser
import threading
import signal
import sys

PORT = 8001
HTML_FILE = "cbtbox.html"

# Start the FastAPI server in a separate thread
def start_fastapi_server():
    subprocess.run(["uvicorn", "microservice:app", "--reload"])

# Open the HTML file in the browser
def open_html_file():
    webbrowser.open_new_tab(f"http://localhost:{PORT}/{HTML_FILE}")

# Start the HTTP server to serve the HTML file
def start_http_server():
    Handler = http.server.SimpleHTTPRequestHandler
    with socketserver.TCPServer(("", PORT), Handler) as httpd:
        print(f"Serving at http://localhost:{PORT}/{HTML_FILE}")
        open_html_file()  # Call to open the HTML file
        httpd.serve_forever()


# Signal handler to capture the interrupt signal (CTRL+C)
def signal_handler(sig, frame):
    print("\nShutting down servers...")
    sys.exit(0)

# Prompt the user for confirmation to stop the servers
def prompt_confirmation():
    confirmation = input("Do you want to stop the servers? (y/n): ")
    if confirmation.lower() == "y":
        print("Shutting down servers...")
        sys.exit(0)


# Start both servers in separate threads
fastapi_thread = threading.Thread(target=start_fastapi_server)
http_thread = threading.Thread(target=start_http_server)

fastapi_thread.start()
http_thread.start()

# Register the signal handler for interrupt signal (CTRL+C)
signal.signal(signal.SIGINT, signal_handler)

# Prompt for user confirmation before exiting
while True:
    prompt_confirmation()

# # Wait for both threads to finish
# fastapi_thread.join()
# http_thread.join()


microservice.py
from typing import Optional
from fastapi import FastAPI, HTTPException
from fastapi.responses import FileResponse
from fastapi.middleware.cors import CORSMiddleware
from gtts import gTTS
import openai
import os
from pymongo import MongoClient
from pymongo.collection import Collection
from pydantic import BaseModel
import pyttsx3

tags_metadata = [
    {
        "name": "chat",
        "description": "You can chat here.",
    },
    {
        "name": "history",
        "description": "History of conversations by the user.",
    },
    {
        "name": "audio",
        "description": "Download audio file of the most recent OpenAI Bot response.",
    },
]

app = FastAPI( openapi_tags=tags_metadata, title="Chatbot Microservice API", description="API to chat with an AI chatbot", version="1.0.0")

origins = ["http://localhost:8001"]  # adjust this to match your actual domain

# Configure CORS middleware
app.add_middleware(
    CORSMiddleware,
    allow_origins=origins,
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)

# Initialize the speech engine
engine = pyttsx3.init()

class ChatInput(BaseModel):
    """Data model for chat input."""
    user_id: str
    message: str
    audio: str = "off"

# Set OpenAI API key
openai.api_key = "***********"  # set your OpenAI API key in environment variables

# Connect to MongoDB client
client = MongoClient(os.getenv("MONGO_CONNECTION_STRING"))
db = client['chatbot']  # Replace with your database name

@app.get("/", include_in_schema=False)
async def root():
    """
    Function to tell server is running whenevr the base URL http://127.0.0.1:8000 is opened.
    """
    return {"message": "Chatbot is running and you can access it at http://127.0.0.1:8000/docs."}

@app.post("/chat", tags=["chat"])
async def chat(chat_input: ChatInput, user_id: str ="Enter your User ID ", message: str ="Chat here", audio: bool = False ) -> dict:
    """
    Function to handle chat with the bot.
    It saves conversation to the database, calls the AI model to get response,
    converts the response to speech and returns the AI response.
    \n **Audio is False = OFF by default, True = ON**
    """
    # user_id = chat_input.user_id
    # message = chat_input.message
    # audio = chat_input.audio
    
    # Load conversation history from DB
    history = db.conversations.find_one({"user_id": user_id})
    if history is None:
        history = {"user_id": user_id, "conversation": []}
    
    # Use OpenAI API to generate a response
    response = openai.ChatCompletion.create(
        model="gpt-3.5-turbo",
        messages=[{
            "role": "system",
            "content": "You are a helpful assistant."
        }, {
            "role": "user",
            "content": message
        }],
        max_tokens=150
    )

    # Check if the response from OpenAI is empty or not
    if response is None or len(response.choices) == 0:
        raise HTTPException(status_code=500, detail="OpenAI response is empty")

    # Store the new messages in the conversation history
    new_messages = {
        "user": message,
        "assistant": response.choices[0].message['content'],
    }
    history['conversation'].append(new_messages)
    db.conversations.update_one({"user_id": user_id}, {"$set": history}, upsert=True)

    # Convert text to speech and speak the response only if audio was requested
    if audio:
            tts = gTTS(text=response.choices[0].message['content'], lang='en')
            tts.save("outputs/output.mp3")

            # Speak the response
            engine.say(response.choices[0].message['content'])
            engine.runAndWait()

    return {"message": response.choices[0].message['content']}  # return the AI response


@app.get("/history/{user_id}", tags=["history"])
async def get_history(user_id: str) -> FileResponse:
    """
    Function to get chat history for a given user id.
    It fetches the conversation from database, converts it to text file and returns as FileResponse.
    """
    history = db.conversations.find_one({"user_id": user_id})
    if history is None:
        raise HTTPException(status_code=404, detail="No conversation history found for this user")
    
    # Convert the conversation list into a single string, each message on a new line
    history_string = "\n".join([f"User: {msg['user']}\nAssistant: {msg['assistant']}" for msg in history["conversation"]])
    
    # Write the history string to a text file
    with open("outputs/history.txt", "w") as file:
        file.write(history_string)

    # Return the file as a response
    return FileResponse("outputs/history.txt", media_type="text/plain", filename=f"{user_id}_history.txt")

@app.get("/audio", tags=["audio"])
async def get_audio() -> FileResponse:
    """
    Function to return the most recent AI response as audio file.
    """
    return FileResponse("outputs/output.mp3", media_type="audio/mpeg", filename="output.mp3")
